server/oauth_service.py

```diff
@@ -0,0 +1,52 @@
+import aiocron
+import aiohttp
+import jwt
+from jwt import InvalidTokenError
+from jwt.algorithms import RSAAlgorithm
+
+from server.config import config
+
+from .core import Service
+from .decorators import with_logger
+from .exceptions import AuthenticationError
+
+
+@with_logger
+class OAuthService(Service, name="oauth_service"):
+    """
+    Service for managing the OAuth token logins and verification.
+    """
+
+    def __init__(self):
+        self.public_keys = None
+
+    async def initialize(self) -> None:
+        await self.retrieve_public_keys()
+        # crontab: min hour day month day_of_week
+        # Run every day to update public keys.
+        self._update_cron = aiocron.crontab(
+            "0 0 * * *", func=self.retrieve_public_keys
+        )
+
+    async def retrieve_public_keys(self) -> None:
+        """
+        Get the latest jwks from the hydra endpoint
+        """
+        self.public_keys = {}
+        async with aiohttp.ClientSession(raise_for_status=True) as session:
+            async with session.get(config.HYDRA_JWKS_URI) as resp:
+                jwks = await resp.json()
```

Comment:
I would only clear the dictionary once all coroutines have been awaited, otherwise you could have a situation where the dictionary is cleared, then while the get request is happening someone tries to log in and can't because there are no keys.

You could also use `dict.clear()` to reuse the same dictionary object instead of creating a new one each time. I don't think anything bad can happen by doing that because of the GIL.

```suggestion
        async with aiohttp.ClientSession(raise_for_status=True) as session:
            async with session.get(config.HYDRA_JWKS_URI) as resp:
                jwks = await resp.json()
                self.public_keys = {}
```

Reply:
I made it so that it should create a new dictionary and then replaces the old dictionary after it is populated so that there should be practically no time with an empty public_keys